### DISCUSSION TITLE
Type Error when used any() and ZodSchema

### DISCUSSION THREAD

**Discussion by Code-Hex** [colinhacks/zod#951] [Q&A]
## What

I wrote some code like below

```typescript
type A = {
  foo: any;
};

function AParser(): z.ZodSchema<A> {
  return z.object({
    foo: z.any()
  });
}
```

## Actual

```
Type 'ZodObject<{ foo: ZodAny; }, "strip", ZodTypeAny, { foo?: any; }, { foo?: any; }>' is not assignable to type 'ZodType<A, ZodTypeDef, A>'.
  Types of property '_type' are incompatible.
    Type '{ foo?: any; }' is not assignable to type 'A'.
      Property 'foo' is optional in type '{ foo?: any; }' but required in type 'A'.
```

## Expected

compilable (no type error)

I understand this is not compilable if I wrote like below

```typescript
type A = {
  foo: any;
};

function AParser(): z.ZodSchema<A> {
  return z.object({
    foo: z.any().optional()
  });
}
```

## Reproduce

https://codesandbox.io/s/zod-example-forked-8qnnq?file=/src/index.ts

**Comment by Code-Hex** [comment]
`z.unknown` has also the same issue.

**Comment by JacobWeisenburger** [comment]
It seems that somewhere in `z.object()`, the type `any` and `unknown` get converted to optional because in typescript, `undefined` extends `any` and `unknown`. see below:
```ts
type test1 = undefined extends any ? true : false
// type test1 = true
type test2 = undefined extends unknown ? true : false
// type test2 = true
```

**Reply by Code-Hex** [reply]
@JacobWeisenburger If so, How do you think that z.any().optional() or nullable() method. I think they are unnecessary. But why exist?

**Reply by JacobWeisenburger** [reply]
They might be unnecessary, but they exist because `ZodAny extends ZodType`

https://github.com/colinhacks/zod/blob/6507e3bd17571ff4003ed20d227bd5864086b106/src/types.ts#L1045

https://github.com/colinhacks/zod/blob/6507e3bd17571ff4003ed20d227bd5864086b106/src/types.ts#L336-L341

**Reply by nahtnam** [reply]
@JacobWeisenburger I completely agree that `.any()` follows how typescript works, but is there any way to exclude `undefined` and/or `null`? Typically you'd use `.required()` but that doesn't exist

**Reply by JacobWeisenburger** [reply]
If you are ok with losing the ability to get the static types from typescript, then using [`refine`](https://github.com/colinhacks/zod#refine) is the best option that I can see.

```ts
const schema = z.object( {
    foo: z.any().refine( input => {
        if ( input != null ) return true
        // this disallows `null` and `undefined`
        // because in `input != null`, `undefined` is coerced to `null`
    } )
} )
```

You might also be able to use `z.custom()`, but I don't see any documentation about this, so I'm not exactly sure it is the same as `z.any().refine()`.

**Comment by lewislwood** [comment]
Just feel bad I had to use the any type.  anyzodobject did not like my schema.

I have validate(schema:AnyZodObject)

I declared a typical userSchema and extend it for regUser schema below
export const regSchema = userSchema .omit({id: true}).extend({
    confirm: z.string()
}).refine((u) => u.confirm === u.password, "Confirmation and Password do not match");

I then call validate(regSchema)   // not valid type of AnyZodObject
